### PR TITLE
Stopping / tracking harvester implementation

### PR DESCRIPTION
**Pull request by ruflin**
The goal of this PR is properly track and shut down harvesters

### CODE REVIEW

**Comment by ruflin** [comment]
@urso This implementation has currently the problem that if the output is not responding, PublishEvent is blocking and filebeat does not stop. Any recommendation on how to handle this case?

**Comment by ruflin** [comment]
This should also get https://github.com/elastic/beats/pull/1148 to green.

File: filebeat/crawler/prospector.go [hunk 1]
```diff
@@ -81,10 +91,13 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
 	// 1. reading from file
 	// 2. forwarding event to spooler
 	// As this is not implemented yet, no blocking on prospector shutdown is done.
-	wg.Done()
 
 	logp.Info("Starting prospector of type: %v", p.ProspectorConfig.Harvester.InputType)
 
+	defer func() {
```
Comment (andrewkroh) [inline filebeat/crawler/prospector.go]: Just use `defer wg.Done()`.

File: filebeat/crawler/prospector.go [hunk 2]
```diff
@@ -98,18 +111,43 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
 }
 
 func (p *Prospector) Stop() {
+	logp.Debug("prospector", "Stopping %v harvesters.", len(p.harvesters))
+	for _, h := range p.harvesters {
+		go h.Stop()
+	}
+	logp.Debug("prospector", "stopp called on all harvesters")
```
Comment (andrewkroh) [inline filebeat/crawler/prospector.go]: `s/stopp/stop/`, but I would prefer to see something like `"Waiting for %d harvesters to stop", len(p.harvesters)`.

File: filebeat/crawler/prospector.go [hunk 3]
```diff
@@ -98,18 +111,43 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
...
+	p.harvestersWaitGroup.Wait()
 	logp.Info("Stopping Prospector")
 	close(p.done)
 }
 
-func (p *Prospector) AddHarvester(file string, stat *harvester.FileStat) (*harvester.Harvester, error) {
+// CreateHarvester creates a harvester based on the give params
```
Comment (andrewkroh) [inline filebeat/crawler/prospector.go]: `s/give/given/`

File: filebeat/crawler/prospector.go (hunk 1 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: Done

File: filebeat/crawler/prospector.go (hunk 2 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: changed

File: filebeat/crawler/prospector.go (hunk 3 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: fixed

File: filebeat/crawler/prospector.go [hunk 4]
```diff
@@ -98,18 +103,43 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
 }
 
 func (p *Prospector) Stop() {
+	logp.Debug("prospector", "Stopping %d harvesters.", len(p.harvesters))
+	for _, h := range p.harvesters {
+		go h.Stop()
+	}
+	logp.Debug("prospector", "Waiting for %d harvesters to stop", len(p.harvesters))
+	p.harvestersWaitGroup.Wait()
```
Comment (urso) [inline filebeat/crawler/prospector.go]: how is p.done used? the prospector loop starting new harvesters must be stopped before shutting down harvesters

File: filebeat/crawler/prospector.go [hunk 5]
```diff
@@ -98,18 +103,43 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
...
+func (p *Prospector) RunHarvester(h *harvester.Harvester) {
+	// Starts harvester and picks the right type. In case type is not set, set it to defeault (log)
+	logp.Debug("harvester", "Starting harvester: %v", h.Id)
+
+	harvesterCounter.Add(1)
+	p.harvestersWaitGroup.Add(1)
+
+	go func() {
+		h.Harvest()
+		p.harvestersWaitGroup.Done()
```
Comment (urso) [inline filebeat/crawler/prospector.go]: prefer defer before calling h.Harvest? What happens with 'Done' call if h.Harvest panics?

File: filebeat/crawler/registrar.go [hunk 6]
```diff
@@ -29,6 +31,7 @@ func NewRegistrar(registryFile string) (*Registrar, error) {
 	r := &Registrar{
 		registryFile: registryFile,
 		done:         make(chan struct{}),
+		wg:           sync.WaitGroup{},
```
Comment (urso) [inline filebeat/crawler/registrar.go]: not required, zero-value already valid.

File: filebeat/harvester/harvester.go [hunk 7]
```diff
@@ -48,11 +53,13 @@ func NewHarvester(
 	}
 
 	h := &Harvester{
+		Id:          uuid.NewV4(), // Unique identifier of each harvester
```
Comment (urso) [inline filebeat/harvester/harvester.go]: why is a random harvester ID required?

File: filebeat/harvester/reader.go [hunk 8]
```diff
@@ -53,6 +55,7 @@ func newLogFileReader(
 		config:       config,
 		lastTimeRead: time.Now(),
 		backoff:      config.backoffDuration,
+		done:         done,
```
Comment (urso) [inline filebeat/harvester/reader.go]: I wonder if channel is required. If file get's closed by another thread (in stop method), we should get some error, right?

**Comment by urso** [comment]
shutdown still seems to be somewhat incomplete (subject to hanging).

once harvester is running, the event flows like this (each step running in it's own go-routine communicating via channels):

```
    harvester -> spooler -> filebeat publisher -> libbeat publisher
                                                                      -> registrar
```

on shutdown ensure the prospector loop looking for new files to be harvested is shutdown before stopping the harvesters.

after stopping harvester, stop the spooler and next the filebeat publisher. the registrar must be stopped last. Not sure if filebeat is stopping it's own publisher worker yet.

when stopping the spooler make, [break the publisher queue at line 148](https://github.com/elastic/beats/blob/master/filebeat/beater/spooler.go#L148).

[The publisher in filebeat](https://github.com/elastic/beats/blob/master/filebeat/beater/publish.go#L14) has a 'Stop' method, but not sure if it's called. Stop on publisher must be called in between spooler and registrar shutdown.

Unfortunately the publisher itself might hang on shutdown in `PublishEvents` to libbeat if logstash/elasticsearch is unresponsive. To solve this, I'm planning to add a `Close` method to libbeat publisher clients, which will break `PublishEvent/PublishEvents` returning an error. This method will be called by the publisher's Stop method in order to break out of unresponsive `PublishEvents`. `Close` being required by all beats will be done in another PR though.

**Comment by urso** [comment]
The publisher worker is initialized and started at [line 96](https://github.com/elastic/beats/blob/master/filebeat/beater/filebeat.go#L96), but never killed.

File: filebeat/harvester/harvester.go (hunk 7 above)
Comment (ruflin) [inline filebeat/harvester/harvester.go]: It's not, see https://github.com/elastic/beats/pull/1144#discussion_r56020188 It is more I quite like UUID.

File: filebeat/crawler/prospector.go (hunk 5 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: done

File: filebeat/crawler/registrar.go (hunk 6 above)
Comment (ruflin) [inline filebeat/crawler/registrar.go]: removed

File: filebeat/harvester/reader.go (hunk 8 above)
Comment (ruflin) [inline filebeat/harvester/reader.go]: So you would suggest to add something like this in log.go before the for loop:

```
	go func() {
		// Closes file so readLine returns error
		// TODO: What happens to this if h.done never closed? Will this be closed if a harvester object is not needed anymore?
		<-h.done
		h.file.Close()
	}()
	for {
		select {
		case <-h.done:
			return
		default:
		}
```

File: filebeat/crawler/prospector.go (hunk 4 above)
Comment (ruflin) [inline filebeat/crawler/prospector.go]: Good point. `close(p.done)` should be moved up before the stopping and it seems like each prospector needs its own waitgroup or channel to notify when all Run methods are stopped.

**Comment by ruflin** [comment]
@urso Most comments applied, except for 

> when stopping the spooler make the publisher break the queue at line 148.

I'm not sure how to break the queue here, as in case it already hangs on line 148, the only thing I know of  to stop it is closing the channel which leads to a panic? I'm sure you had a better way in mind :-)

**Comment by tsg** [comment]
Waiting for #1402.

File: filebeat/harvester/reader.go (hunk 8 above)
Comment (ruflin) [inline filebeat/harvester/reader.go]: @urso I tried this now in my new version and it actually leads to a race condition ...

**Comment by ruflin** [comment]
Closing as this one got replaced by https://github.com/elastic/beats/pull/1604

**Pull request closed without merging**